programs/host/frontend.c

```diff
@@ -0,0 +1,937 @@
+
+    switch( function )
+    {
+        case MBEDTLS_SERIALIZE_FUNCTION_EXIT:
+            ret = 0;
+            ctx_p->status = MBEDTLS_SERIALIZE_STATUS_EXITED;
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_ECHO:
+            {
+                CHECK_ARITY( 1 );
+                ALLOC_OUTPUT( 0, inputs[0]->size );
+                DBG( "executing echo" );
+                memcpy( item_buffer( outputs[0] ), item_buffer( inputs[0] ),
+                        inputs[0]->size );
+                ret = 0;
+                break;
+            }
+
+        case MBEDTLS_SERIALIZE_FUNCTION_USLEEP:
+            CHECK_ARITY( 1 );
+            CHECK_LENGTH( 0, 4 ); // usec
+            {
+                unsigned long usec = item_uint32( inputs[0] );
+                DBG( "executing sleep usec=%lu", usec );
+                mbedtls_net_usleep( usec );
+                ret = 0;
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_SOCKET:
+            CHECK_ARITY( 3 ); // host, port, proto_and_mode
+            CHECK_LENGTH( 2, 2 ); // proto_and_mode
+            {
+                char *host = item_buffer( inputs[0] );
+                char *port = item_buffer( inputs[1] );
+                if ( host[inputs[0]->size - 1] == '\0' && port[inputs[1]->size - 1] == '\0' )
+                {
+                    uint16_t proto = item_uint16( inputs[2] );
+                    int is_bind =
+                        ( ( proto & MBEDTLS_SERIALIZE_SOCKET_DIRECTION_MASK ) ==
+                          MBEDTLS_SERIALIZE_SOCKET_BIND );
+                    mbedtls_net_context net_ctx;
+                    ALLOC_OUTPUT( 0, 2 ); // fd
+                    proto &= ~MBEDTLS_SERIALIZE_SOCKET_DIRECTION_MASK;
+                    if( is_bind )
+                    {
+                        DBG( "executing socket/bind" );
+                        ret = mbedtls_net_bind( &net_ctx, host, port, proto );
+                    }
+                    else
+                    {
+                        DBG( "executing socket/connect" );
+                        ret = mbedtls_net_connect( &net_ctx, host, port, proto );
+                    }
+                    if( ret == 0 )
+                    {
+                        DBG( "socket -> fd %d", (int)net_ctx.fd );
+                        set_item_uint16( outputs[0], net_ctx.fd );
+                    }
+                }
+                else
+                {
+                    DBG( "host and/or port string not null terminated!" );
+                    ret = MBEDTLS_ERR_SERIALIZE_BAD_INPUT;
+                }
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_ACCEPT:
+            CHECK_ARITY( 2 );
+            CHECK_LENGTH( 0, 2 ); // socket_fd
+            CHECK_LENGTH( 1, 4 ); // buffer_size
+            {
+                mbedtls_net_context bind_ctx = { item_uint16( inputs[0] ) };
+                uint32_t buffer_size = item_uint32( inputs[1] );
```

Comment:
Is there a limit we might want to set here before allocating something of size `buffer_size`?

Reply:
It is the size of the buffer to hold the peer IP address. Limit wise it is MAX_INT. But Xalloc would fail for any faulty large value. I don't think we need to check mem allocation requests.